Fix #10037: Adds a customization arg for math interactions.

iamprayush opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #10037.
2. This PR does the following: Adds a customization arg that allows creators to specify the letters that the learner will see in the on-screen keyboard.
This is how the arg behaves:
![custarg](https://user-images.githubusercontent.com/35144226/89105981-93746e80-d443-11ea-9822-f62a169849c3.gif)


<details>
  <summary>Since this PR adds a new customization arg along with e2e tests, here are the success runs of Circle and Travis CI:</summary>
 
Travis 1:
![travis1](https://user-images.githubusercontent.com/35144226/89740770-55152a00-daa9-11ea-8249-4b46019caf66.png)
Travis 2:
![travis2](https://user-images.githubusercontent.com/35144226/89740771-55adc080-daa9-11ea-9498-aa7aaa13e8c7.png)
Travis 3:
![travis3](https://user-images.githubusercontent.com/35144226/89740773-56465700-daa9-11ea-8bab-d598b1f61a14.png)

Circle 1:
![circle1](https://user-images.githubusercontent.com/35144226/89740766-521a3980-daa9-11ea-9cb6-f1c78ac36b22.png)
Circle 2:
![circle2](https://user-images.githubusercontent.com/35144226/89740768-53e3fd00-daa9-11ea-9c0d-21ff7f9ab58c.png)
Circle 3:
![circle3](https://user-images.githubusercontent.com/35144226/89740769-547c9380-daa9-11ea-9778-8a222adf88bc.png)


</details>

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

iamprayush committed:
Initial commit.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Fully implemented the customization arg.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Added current letters display.

iamprayush committed:
Added state migration.

iamprayush committed:
Minor fix

iamprayush committed:
Added tests for question domain fn.

iamprayush committed:
Added frontend tests.

iamprayush committed:
Fixed linter errors.

iamprayush committed:
Fixed linter errors.

iamprayush committed:
Fixed linter errors.

iamprayush committed:
Fixed linter errors.

iamprayush commented:
/cc @mschanteltc @rachelwchen

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Added OSK customization based on cust arg and type of interaction.

mschanteltc commented:
Great work Prayush :) A few comments/questions:

- Are we going to enforce a maximum limit? If so, can we add something similar to a [character counter](https://www.google.com/search?q=character+counter+ui&hl=en&sxsrf=ALeKk03JpgHEodbRdWlCzptaGgSt6mOPpw:1596487552788&source=lnms&tbm=isch&sa=X&ved=2ahUKEwinkePL8__qAhWKCTQIHZTvCsYQ_AUoAnoECA8QBA&biw=1920&bih=977#imgrc=RXGh2DLtzW8aKM) that decreases as variables are added? We can say something like "x variables remaining"

- As variables are being added, does the container with the selected variables reflect how these variables will be ordered from the learner's view? If so, we should specify that in the description such as "The order of these variables will be reflected in the learner's keyboard." If not, we might need to discuss how we would prioritize them. 

- I think it might be useful to disable some capital greek letters because of how identical they appear to the Latin alphabet (Alpha -> A, Beta -> B, Eta -> H). I think it would be safe to assume that some of these capitalized Greek letters would be unnecessary in any topic. @seanlip WDYT?

seanlip commented:
I think disabling the selection of such greek letters entirely is reasonable. But @mschanteltc, what do you mean by "I think it would be safe to assume that some of these capitalized Greek letters would be necessary in any topic"? I would have thought that they would be completely unnecessary.

mschanteltc commented:
> I think disabling the selection of such greek letters entirely is reasonable. But @mschanteltc, what do you mean by "I think it would be safe to assume that some of these capitalized Greek letters would be necessary in any topic"? I would have thought that they would be completely unnecessary.

Sorry, I meant "unnecessary." I corrected my comment to fix this misunderstanding.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Fixed backend test failure.

iamprayush committed:
Fixed e2e tests

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Modified description, added character counter, and removed some capital greek letters.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Fixed linter errors.

iamprayush commented:
Thanks for the comments, @mschanteltc!

> * Are we going to enforce a maximum limit? If so, can we add something similar to a character counter that decreases as variables are added? We can say something like "x variables remaining"

Yes, I have added a validation hat limits the number of letters selected to 10. Currently, it just shows a warning and disables the save button if more than 10 letters are selected. I have now also added the character count to further assist the creator. It looks like this:
![validation](https://user-images.githubusercontent.com/35144226/89319191-6b804780-d69d-11ea-8774-a10547a14e7b.gif)


> * As variables are being added, does the container with the selected variables reflect how these variables will be ordered from the learner's view? If so, we should specify that in the description such as "The order of these variables will be reflected in the learner's keyboard." If not, we might need to discuss how we would prioritize them.

Yes, the order will be the same. I have updated the description accordingly.

> * I think it might be useful to disable some capital greek letters because of how identical they appear to the Latin alphabet (Alpha -> A, Beta -> B, Eta -> H). I think it would be safe to assume that some of these capitalized Greek letters would be unnecessary in any topic. @seanlip WDYT?

I agree. I have removed the following capital Greek letters:
`Alpha (Α), Beta (Β), Epsilon (Ε), Zeta (Ζ), Eta (Η), Iota (Ι), Kappa (Κ), Mu (Μ), Nu (Ν), Rho (Ρ), Tau (Τ), Upsilon (Υ), Chi (Χ)`

iamprayush committed:
Fixed backend failures

iamprayush committed:
Removed guppy OSK stuff.

aks681 commented on core/domain/exp_domain.py:
```diff
@@ -2734,6 +2734,58 @@ def _convert_states_v34_dict_to_v35_dict(cls, states_dict):
+        Returns:
+            dict. The converted states_dict.
+        """
+        greek_symbols = constants.GREEK_SYMBOLS_LOWERCASE + (
+            constants.GREEK_SYMBOLS_UPPERCASE)
+        greek_letters = constants.GREEK_LETTERS
```
What's the difference? Is one the display value and one the input value?

iamprayush replied:
`GREEK_LETTERS` represent the words and `GREEK_SYMBOLS` represent the symbols. I thought naming it to something like `GREEK_WORDS` seems weird which is why I chose to name it this way.

seanlip replied:
I suggest going with GREEK_LETTER_NAMES -- that would be less confusing, IMO.

aks681 commented on core/domain/exp_domain.py:
```diff
@@ -2734,6 +2734,58 @@ def _convert_states_v34_dict_to_v35_dict(cls, states_dict):
+                            if greek_letter in rule_input:
+                                rule_input = rule_input.replace(
+                                    greek_letter, '')
+                                variables.add(greek_symbols[ind])
```
Does this mean the constants (in constants.ts) should have them in the same order? If so, maybe add a comment there?

iamprayush replied:
Yes, I've added a comment.

aks681 commented on core/domain/exp_domain.py:
```diff
@@ -2734,6 +2734,58 @@ def _convert_states_v34_dict_to_v35_dict(cls, states_dict):
+                        # ['β', 'b', 'e', 't', 'a', 'x'].
+                        for ind, greek_letter in enumerate(greek_letters):
+                            if greek_letter in rule_input:
+                                rule_input = rule_input.replace(
+                                    greek_letter, '')
```
Just to check, this won't modify the rule in the input, right?

iamprayush replied:
Yes, it won't modify the rule inputs.

aks681 commented on core/domain/exp_domain.py:
```diff
@@ -2734,6 +2734,58 @@ def _convert_states_v34_dict_to_v35_dict(cls, states_dict):
+                customization_args = state_dict[
+                    'interaction']['customization_args']
+                customization_args.update({
+                    'customOskLetters': {
```
customOSKLetters?

iamprayush replied:
That won't work since `customOSKLetters` would get converted to `custom-oskletters-editor` instead of `custom-osk-letters-editor`.

aks681 commented on core/domain/question_domain.py:
```diff
@@ -459,6 +459,58 @@ def _convert_state_v34_dict_to_v35_dict(cls, question_state_dict):
         return question_state_dict
 
+    @classmethod
+    def _convert_state_v35_dict_to_v36_dict(cls, question_state_dict):
```
Ditto comments in state_domain here also.

iamprayush replied:
Addressed above.

aks681 commented on core/templates/services/math-interactions.service.ts:
```diff
@@ -184,6 +184,35 @@ export class MathInteractionsService {
   }
 
   insertMultiplicationSigns(expressionString: string): string {
```
Why are the changes below in `insertMultiplicationSigns` function? Also, why are the changes below required?
Since, it was working without these before, right?

iamprayush replied:
The problem was that the older implementation didn't account for consecutive greek words without multiplication signs so things like `alphabeta` would be considered as `a*l*p*h*a*b*e*t*a` instead of `alpha*beta`. With these changes, such cases are also covered. Also, cases like `xalphayzgamma` are converted to `x*alpha*y*z*gamma`.

aks681 commented on extensions/objects/models/objects.py:
```diff
@@ -1146,3 +1146,25 @@ class PositionOfTerms(BaseObject):
+
+
+class CustomOskLetters(BaseObject):
+    """Class fot custom OSK letters. These are the letters that will be
```
typo: 'fot'

iamprayush replied:
Fixed

aks681 commented on extensions/objects/models/objects.py:
```diff
@@ -1146,3 +1146,25 @@ class PositionOfTerms(BaseObject):
+class CustomOskLetters(BaseObject):
+    """Class fot custom OSK letters. These are the letters that will be
+    displayed to the learner for AlgebraicExpressionInput and MathEquationInput
+    interactions when the on-screen keyboard is being used. This includes roman
```
Do you mean 'roman' or 'latin'? Or are they the same :P?

iamprayush replied:
Apparently they're the same ([ref](https://en.wikipedia.org/wiki/Latin_alphabet)), but just to be consistent, I've changed this to Latin, since I've used Latin everywhere else.

aks681 reviewed (commented):
Code looks good overall! 
Haven't run it locally yet, will comment after that as well.

mschanteltc commented:
Thanks for the updates @iamprayush! The user experience of this looks great so far. Here are some comments I have:

- Let's use a regular weight on the description "Short variables... keyboard)" 
- For "You can still add x more letter(s)" right align it to the selected variables container. Decrease the font size and use regular size font. 
- Would it be possible to have an empty container when no letters are selected? That way nothing below it shifts down once the first letter is selected. Let's use a light gray border like the borders used in the unselected letter buttons.

iamprayush committed:
Resolved review comments.

iamprayush commented:
> * Let's use a regular weight on the description "Short variables... keyboard)"

That description field is actually common between all the interactions. So, the styles for that would apply to all the other interactions as well. Should I still change it to regular weight?

> * For "You can still add x more letter(s)" right align it to the selected variables container. Decrease the font size and use regular size font.

Done.

> * Would it be possible to have an empty container when no letters are selected? That way nothing below it shifts down once the first letter is selected. Let's use a light gray border like the borders used in the unselected letter buttons.

Done.

This is how it looks now:
![new](https://user-images.githubusercontent.com/35144226/89354353-40feb080-d6d6-11ea-9e61-02f1181b5b9f.gif)

mschanteltc commented:
> > * Let's use a regular weight on the description "Short variables... keyboard)"
> 
> That description field is actually common between all the interactions. So, the styles for that would apply to all the other interactions as well. Should I still change it to regular weight?
> 
I believe that after the style guide is developed, we can apply all style changes to headings, so for now, leave it as is. 
> > * For "You can still add x more letter(s)" right align it to the selected variables container. Decrease the font size and use regular size font.
> 
> Done.
> 
> > * Would it be possible to have an empty container when no letters are selected? That way nothing below it shifts down once the first letter is selected. Let's use a light gray border like the borders used in the unselected letter buttons.
> 
> Done.
> 
> This is how it looks now:
> ![new](https://user-images.githubusercontent.com/35144226/89354353-40feb080-d6d6-11ea-9e61-02f1181b5b9f.gif)

Awesome! And last thing, are creators who are using this on a web browser able to add letters by typing on their keyboard? If so, are creators able to delete the last character they typed in by pressing the 'backspace' button? If this functionality does not exist, @seanlip do you think there is merit to adding something like this?

iamprayush commented:
> And last thing, are creators who are using this on a web browser able to add letters by typing on their keyboard? If so, are creators able to delete the last character they typed in by pressing the 'backspace' button?

No, even on desktops, the user would be required to click the buttons that are provided. Personally, I think that the keyboard input functionality is unnecessary but if that's something that is required, I can implement it (along with the backspace button functionality). Although note that, it might be confusing to add greek letters if we allow keyboard input.

/cc @aks681

seanlip commented:
> No, even on desktops, the user would be required to click the buttons that are provided. Personally, I think that the keyboard input functionality is unnecessary but if that's something that is required, I can implement it (along with the backspace button functionality). Although note that, it might be confusing to add greek letters if we allow keyboard input.

I think keyboard input functionality on desktop is necessary, actually. It's much faster to type a simple expression than to hunt-and-peck with the mouse. You don't need to support typing greek letters on the keyboard, but you should know the effect of pressing each key on the keyboard (some keys may do nothing). Also, as a bonus, I would suggest highlighting/flashing the key in the on-screen keyboard when the corresponding physical keyboard key is pressed (switching tabs if necessary).

(Also, I think it's fine to do this in a separate PR btw -- just make sure to file an issue for it. Thanks!)

iamprayush committed:
Fixed merge conflicts.

iamprayush committed:
Resolved merge conflicts.

iamprayush committed:
Fixed linter errors.

iamprayush committed:
Added keyboard event functionality

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Fixed linter errors.

iamprayush commented:
> I think keyboard input functionality on desktop is necessary, actually. It's much faster to type a simple expression than to hunt-and-peck with the mouse. You don't need to support typing greek letters on the keyboard, but you should know the effect of pressing each key on the keyboard (some keys may do nothing). Also, as a bonus, I would suggest highlighting/flashing the key in the on-screen keyboard when the corresponding physical keyboard key is pressed (switching tabs if necessary).

@seanlip I've implemented the keyboard event listeners and this is how it functions now on desktop:
![keyboards](https://user-images.githubusercontent.com/35144226/89534032-f6824e80-d811-11ea-88be-5c16ad9c0ff8.gif)

iamprayush commented:
Hey @vojtechjelinek @seanlip @kevintab95 @nithusha21 @DubeySandeep @U8NWXD PTAL as code-owners, thanks!

iamprayush committed:
Minor change.

vojtechjelinek reviewed (approved):
LGTM!

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

nithusha21 commented on extensions/interactions/AlgebraicExpressionInput/directives/algebraic-expression-input-validation.service.spec.ts:
```diff
@@ -184,4 +188,52 @@ describe('AlgebraicExpressionInputValidationService', () => {
+      type: AppConstants.WARNING_TYPES.ERROR,
+      message: (
+        'The following variables are present in some of the answer groups ' +
+        'but are missing from the custom letters list: α,x')
```
Is this a learner facing error message? If so, how does a learner know what the "custom letters list" is?

iamprayush replied:
No, this is a creator facing message and the creator has the ability to modify the custom letters and the rules.

nithusha21 commented on extensions/interactions/AlgebraicExpressionInput/directives/algebraic-expression-input-validation.service.spec.ts:
```diff
@@ -184,4 +188,52 @@ describe('AlgebraicExpressionInputValidationService', () => {
+      currentState, customizationArgs, answerGroups, goodDefaultOutcome);
+    expect(warnings).toEqual([{
+      type: AppConstants.WARNING_TYPES.ERROR,
+      message: 'The number of custom letters cannot be more than 10.'
```
Same question as above

iamprayush replied:
Addressed.

nithusha21 commented on extensions/interactions/MathEquationInput/directives/math-equation-input-validation.service.spec.ts:
```diff
@@ -188,4 +192,52 @@ describe('MathEquationInputValidationService', () => {
+    expect(warnings).toEqual([{
+      type: AppConstants.WARNING_TYPES.ERROR,
+      message: (
+        'The following variables are present in some of the answer groups ' +
```
Ditto

iamprayush replied:
Addressed

nithusha21 reviewed (approved):
Just some questions

U8NWXD commented on extensions/interactions/AlgebraicExpressionInput/protractor.js:
```diff
@@ -19,8 +19,17 @@
-  // There are no customizations.
+var customizeInteraction = async function(elem, customLetters) {
+  expect(await elem.element(
+    by.css('.protractor-test-custom-letters-div')).isPresent()).toBe(true);
```
Can you make this a waitFor statement instead? It will do essentially the same thing (check that the element is present) but wait a little bit for the element to load before failing

iamprayush replied:
Done.

U8NWXD commented on extensions/interactions/AlgebraicExpressionInput/protractor.js:
```diff
@@ -19,8 +19,17 @@
+    by.css('.protractor-test-custom-letters-div')).isPresent()).toBe(true);
+  for (let letter of customLetters) {
+    await elem.element(by.buttonText('abc')).click();
+    let letterIsPresent = await elem.element(by.buttonText(letter)).isPresent();
```
This check also doesn't allow for the element to still be loading when this line runs. You can check this with a `try` like this:

```js
let letterIsPresent = true;
try {
  waitFor.presenceOf(element, "This error should be caught.");
} catch (e) {
  letterIsPresent = false;
}
```

iamprayush replied:
Alright, so I tried calling the presenseOf function as you've suggested, but the error doesn't seem to get caught for some reason. I tried explicitly calling the function's statements directly and it works. Is that fine?

U8NWXD replied:
Could you dig a little into why just calling the function doesn't work? I don't see why it would matter whether we call the function or its lines, and we should avoid copying function code instead of calling the function

U8NWXD commented on extensions/interactions/AlgebraicExpressionInput/protractor.js:
```diff
@@ -19,8 +19,17 @@
+  expect(await elem.element(
+    by.css('.protractor-test-custom-letters-div')).isPresent()).toBe(true);
+  for (let letter of customLetters) {
+    await elem.element(by.buttonText('abc')).click();
```
Please make sure the element has loaded before trying to click it. You can use `protractor_utils/action.js` to do this easily

U8NWXD replied:
This is a common problem we see in e2e test PR reviews. You can use [this checklist](https://github.com/oppia/oppia/wiki/End-to-End-Tests#codeowner-checks) to check for these common problems when you're writing PRs

iamprayush replied:
I understand, I've fixed this now and I will keep this in mind for future, thanks!

U8NWXD commented on extensions/interactions/AlgebraicExpressionInput/protractor.js:
```diff
@@ -19,8 +19,17 @@
+    await elem.element(by.buttonText('abc')).click();
+    let letterIsPresent = await elem.element(by.buttonText(letter)).isPresent();
+    if (!letterIsPresent) {
+      await elem.element(by.buttonText('αβγ')).click();
+    }
+    await elem.element(by.buttonText(letter)).click();
```
ditto: check elements have loaded before clicking

iamprayush replied:
Done.

U8NWXD commented on extensions/interactions/MathEquationInput/protractor.js:
```diff
@@ -19,8 +19,17 @@
 
-var customizeInteraction = function() {
-  // There are no customizations.
+var customizeInteraction = async function(elem, customLetters) {
+  expect(await elem.element(
+    by.css('.protractor-test-custom-letters-div')).isPresent()).toBe(true);
+  for (let letter of customLetters) {
+    await elem.element(by.buttonText('abc')).click();
+    let letterIsPresent = await elem.element(by.buttonText(letter)).isPresent();
+    if (!letterIsPresent) {
+      await elem.element(by.buttonText('αβγ')).click();
+    }
+    await elem.element(by.buttonText(letter)).click();
+  }
```
Same comments as I left for AlgebraicExpressionInput:

* Use waitFor statements instead of `.isPresent()` to allow for elements needing to load
* Check elements have loaded before clicking them

iamprayush replied:
Done.

U8NWXD reviewed (changes requested):
@iamprayush I left some comments. Since this PR changes the interaction utils that are used across lots of tests, please post screenshots of 3 consecutive passes of the entire test set. Details in the wiki: https://github.com/oppia/oppia/wiki/End-to-End-Tests#temporary-flakiness-mitigation-measures

kevintab95 reviewed (approved):
LGTM for codeowner files. Thanks @iamprayush!

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Fixed backend test failures and a minor bug.

iamprayush committed:
Fixed scrolling issue in OSK.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Resolved e2e test comments.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Fixed linter errors.

iamprayush committed:
Fixed linter errors.

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -2829,111 +2951,41 @@ def _convert_states_v34_dict_to_v35_dict(cls, states_dict):
+                            if greek_letter in rule_input:
+                                rule_input = rule_input.replace(
+                                    greek_letter, '')
+                                variables.add(greek_symbols[ind])
```
This sort of thing suggests that the data structure is wrong. Can you bundle them into a single list instead (or a dict if the order does not matter), so that the elements are automatically paired up?

iamprayush replied:
Done. I've combined them into a dict.

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -2829,111 +2951,41 @@ def _convert_states_v34_dict_to_v35_dict(cls, states_dict):
+                        # ['β', 'b', 'e', 't', 'a', 'x'].
+                        for ind, greek_letter in enumerate(greek_letters):
+                            if greek_letter in rule_input:
+                                rule_input = rule_input.replace(
```
The logic seems too ad hoc: does it handle stuff like `betalphaa` correctly? Can you prove that it is equivalent to your parser output?

It would be better to parse this using your parser and then identify the variables etc. from there.

iamprayush replied:
Done. I've added a function in the parser file for extracting variables.

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -2829,111 +2951,41 @@ def _convert_states_v34_dict_to_v35_dict(cls, states_dict):
+                    'interaction']['customization_args']
+                customization_args.update({
+                    'customOskLetters': {
+                        'value': list(variables)
```
Sort them first? It's a bit cleaner.

iamprayush replied:
Done.

seanlip reviewed (changes requested):
Left a few notes -- only took a look at two codeowner files feconf.py and schema_utils_test.py, and a glance at the migration logic itself.

The migration needs to be pre-run though. Could you write an audit job that does the migration on all exps (without saving) and fill in the form to run it? Have that job output the result of the expressions converted to letters, so that we can do a spot-check on the results to make sure that the conversion is correct.

seanlip commented:
@iamprayush -- re my top-level comment, please see #10186, since I think it might be useful for you and save you quite a bit of hassle!

If you like, once you merge it into your branch, you could tackle the follow-up fixes, but let @shavavo know if you're doing that so that there isn't unnecessary overlap. Thanks!

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Refactored greek letters storage.

iamprayush committed:
Modified general audit job output.

iamprayush committed:
Modified migration audit job for math output.

iamprayush committed:
Fixed linter errors.

iamprayush committed:
Fixed linter errors.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Fixed frontend failures.

iamprayush commented:
> @iamprayush -- re my top-level comment, please see #10186, since I think it might be useful for you and save you quite a bit of hassle!
> 
> If you like, once you merge it into your branch, you could tackle the follow-up fixes, but let @shavavo know if you're doing that so that there isn't unnecessary overlap. Thanks!

@seanlip, I've temporarily modified @shavavo 's general audit just to output the extracted variables for the manual check as you suggested. PTAL!

iamprayush commented:
@DubeySandeep @U8NWXD @aks681 PTAL as code-owners, thanks!

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

U8NWXD commented on extensions/interactions/AlgebraicExpressionInput/protractor.js:
```diff
@@ -17,10 +17,29 @@
-  // There are no customizations.
+var customizeInteraction = async function(elem, customLetters) {
+  await waitFor.presenceOf(elem.element(by.css(
+    '.protractor-test-custom-letters-div')), 'Element took too long to load.');
```
Can you make this error message more specific? In particular, describe which element took long to load. This is important because we don't always get reliable line numbers in stack traces, so the message has to be enough to debug any problems

iamprayush replied:
Done.

U8NWXD commented on extensions/interactions/AlgebraicExpressionInput/protractor.js:
```diff
@@ -17,10 +17,29 @@
+    await action.click('Math OSK Tab', elem.element(by.buttonText('abc')));
+    let letterIsPresent = true;
+    try {
+      await browser.wait(
+        await waitFor.until.presenceOf(elem.element(by.buttonText(letter))),
+        waitFor.DEFAULT_WAIT_TIME_MSECS, 'Element took too long to load.');
```
Could you dig into why just calling the waitFor function doesn't work? Copying the function code is really not ideal

iamprayush replied:
I did some more experimenting and it seems like calling the function doesn't explicitly throw an exception which is why the catch block is never being reached. Now, I tried adding a `throw new Error(errorMessage);` in the `presenceOf` function and then it threw exceptions. But, having this would mean that we would have to wrap all invocations of the `presenceOf` function in a try-catch block, otherwise, they will throw errors in the absence of said element. Also, I don't see such a pattern of catching errors from waitFor functions elsewhere in the codebase.

iamprayush replied:
Update: I've used a much simpler approach (thanks @aks681!), which makes this problem obsolete.

U8NWXD replied:
@iamprayush Here's an example where we use a similar pattern:

https://github.com/oppia/oppia/blob/develop/core/tests/protractor_utils/TopicsAndSkillsDashboardPage.js#L342

I'll also dig into this some more on my end to see if I can figure out why it isn't working for you

iamprayush replied:
I see. I tried this pattern (I also called the click function inside the try block and it worked).
```javascript
for (let letter of customLetters) {
    await action.click('Math OSK Tab', elem.element(by.buttonText('abc')));
    try {
      await waitFor.presenceOf(
        elem.element(by.buttonText(letter)),
        'Letter took too long to be present.');
      await elem.element(by.buttonText(letter)).click();
    } catch (e) {
      await action.click('Math OSK Tab', elem.element(by.buttonText('αβγ')));
    }
    await action.click('Math OSK Letter', elem.element(by.buttonText(letter)));
  }
```

U8NWXD reviewed (changes requested):
@iamprayush PTAL. Also don't forget the screenshots of 3 consecutive passes of the entire set of e2e tests

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Fixed backend failures.

iamprayush committed:
e2e error change.

iamprayush committed:
Fixed linter errors.

seanlip commented on assets/constants.ts:
```diff
@@ -5307,12 +5307,45 @@ export = {
+
+  // Number of custom letters allowed in the on-screen keyboard for math
+  // interactions.
+  "ALLOWED_CUSTOM_LETTERS_LIMIT": 10,
```
MAX_CUSTOM_LETTERS_FOR_OSK

(Use more specific names. The current name doesn't specify what the limit is for. It doesn't specify whether it's an upper/lower limit. When naming things, try to be specific.)

iamprayush replied:
Done.

seanlip commented on extensions/objects/templates/custom-osk-letters-editor.component.ts:
```diff
@@ -0,0 +1,102 @@
+          ctrl.latinLowerCase.join('').indexOf(
+            keyPressed.toLowerCase()) !== -1 &&
+          ctrl.value.indexOf(keyPressed) === -1) {
+          $('button#custom-letter-' + keyPressed).click();
```
No please don't add new jQuery constructs, here and elsewhere. Find a way to call the actual function.

iamprayush replied:
Done.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Minor fixes

seanlip reviewed (approved):
LGTM. Thanks!

iamprayush committed:
Fixed audit job output.

iamprayush committed:
Fixed audit job output.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -325,7 +325,10 @@ def map(item):
     def reduce(key, values):
         if key == 'SUCCESS':
-            yield (key, values[:VALID_MATH_INPUTS_YIELD_LIMIT])
+            if values[0] == 'None':
```
No -- I mentioned this already in chat. Use a different key for this case.

iamprayush replied:
Done.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -325,7 +325,10 @@ def map(item):
+            if values[0] == 'None':
+                yield (key, len(values))
+            else:
+                yield (key, values[:VALID_MATH_INPUTS_YIELD_LIMIT])
```
Why do you still have a limit? Don't do that.

iamprayush replied:
I've removed it now. I thought we just need to see a sample of the output to verify manually.

seanlip commented:
Please also fix the job to migrate any version to N-1.

iamprayush committed:
Modified audit job to accomodate older explorations.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush commented:
> Please also fix the job to migrate any version to N-1.

I've added this now as well. Now, any exploration that has version less than N-1 will first be migrated to N-1 and then the audit job will be run.

iamprayush committed:
Minor changes in e2e tests

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush commented:
@U8NWXD @DubeySandeep PTAL, thanks!

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -253,8 +253,8 @@ def reduce(key, values):
 class ExplorationMigrationAuditJob(jobs.BaseMapReduceOneOffJobManager):
-    """A reusbale one-off job for testing exploration migration from the
-    previous exploration schema version to the latest. This job runs the state
+    """A reusbale one-off job for testing exploration migration from any
```
nit: reusable

iamprayush replied:
Done.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -273,62 +273,72 @@ def map(item):
+                'to_version': python_utils.UNICODE(
+                    feconf.CURRENT_STATE_SCHEMA_VERSION)
+            })]
+            exp_services.update_exploration(
```
No -- don't make any changes to the datastore. This is an _audit_ job. Do the migration in place using the conversion_fn.

iamprayush replied:
I understand. I've made changes so that there are no changes made to the datastore.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -273,62 +273,72 @@ def map(item):
+        # Upgrading older explorations to current version - 1 before running the
+        # audit job.
+        if item.states_schema_version < current_state_schema_version - 1:
+            feconf.CURRENT_STATE_SCHEMA_VERSION -= 1
```
Do not modify constants.

iamprayush replied:
Done.

seanlip commented on core/tests/protractor.conf.js:
```diff
@@ -242,6 +242,7 @@ exports.config = {
       // this thread: https://github.com/angular/protractor/issues/5274
       w3c: false,
       args: [
+        '--headless',
```
Is the e2e team ok with this? Seems like the sort of change that should be run past them. /cc @kevintab95 @nithusha21 @U8NWXD @DubeySandeep

iamprayush replied:
Oh, this is unintentional. This was just for testing locally, I've reverted it now.

seanlip reviewed (changes requested):
@iamprayush The job still cannot be run. It breaks assumptions of audit jobs. PTAL at the comments.

iamprayush committed:
Fixed nits.

iamprayush committed:
Resolved merge conflicts.

iamprayush committed:
Minor change

iamprayush committed:
Minor change

iamprayush committed:
Minor fix

iamprayush committed:
Modified audit job.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Fixed linter errors.

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -3985,16 +4009,15 @@ def _convert_v40_dict_to_v41_dict(cls, exploration_dict):
         Args:
             exploration_dict: dict. The dict representation of an exploration
-                with schema version v41.
+                with schema version v39.
```
The numbers here seem wrong?

iamprayush replied:
Fixed.

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -3985,16 +4009,15 @@ def _convert_v40_dict_to_v41_dict(cls, exploration_dict):
         Returns:
             dict. The dict representation of the Exploration domain object,
-            following schema version v42.
+            following schema version v41.
```
Ditto

iamprayush replied:
Fixed.

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -253,8 +253,8 @@ def reduce(key, values):
 class ExplorationMigrationAuditJob(jobs.BaseMapReduceOneOffJobManager):
-    """A reusable one-off job for testing exploration migration from the
-    previous exploration schema version to the latest. This job runs the state
+    """A reusable one-off job for testing exploration migration from any
```
No, this test still tests migration of the previous exploration schema version to the latest one. (The setup involves bringing the exp to N-1, but the job itself is testing out the N-1 --> N migration.)

iamprayush replied:
Changed it now.

seanlip commented on core/domain/exp_jobs_one_off_test.py:
```diff
@@ -1133,16 +1127,16 @@ def test_migration_job_audit_success(self):
 
         self.create_exploration_with_states_schema_version(
-            36,
+            37,
```
Why do you need to change all these? The idea was to fix the start and end versions so that this doesn't need to be changed each time someone wants to do a migration.

iamprayush replied:
I see. I've reverted these changes now.

seanlip commented on core/domain/exp_jobs_one_off_test.py:
```diff
@@ -1223,7 +1317,7 @@ def test_migration_job_audit_failure(self):
 
         self.create_exploration_with_states_schema_version(
-            36,
+            37,
```
Ditto.

iamprayush replied:
Fixed.

iamprayush committed:
Modified audit job.

iamprayush committed:
Reverted change.

iamprayush committed:
Fixed linter errors.

seanlip commented:
@DubeySandeep @U8NWXD @aks681 PTAL as codeowners. Thanks!

DubeySandeep reviewed (approved):
LGTM for the codeowner files.

iamprayush committed:
Merge branch 'develop' of https://github.com/oppia/oppia into osk-cust-arg

iamprayush committed:
Reverted success output changes.

aks681 reviewed (approved):
Lgtm!

iamprayush committed:
Modified get_variables function

U8NWXD commented on core/tests/protractor_utils/waitFor.js:
```diff
@@ -172,3 +172,5 @@ exports.invisibilityOfInfoToast = invisibilityOfInfoToast;
 exports.visibilityOfInfoToast = visibilityOfInfoToast;
 exports.visibilityOfSuccessToast = visibilityOfSuccessToast;
 exports.modalPopupToAppear = modalPopupToAppear;
+exports.until = until;
+exports.DEFAULT_WAIT_TIME_MSECS = DEFAULT_WAIT_TIME_MSECS;
```
We don't need to export these anymore do we?

iamprayush replied:
Yup, removed.

U8NWXD reviewed (changes requested):
@iamprayush I just have one last comment, but it can be addressed in a follow-up PR. I'll approve once you post screenshots of 3 consecutive passes of the entire CircleCI and TravisCI test sets

U8NWXD commented:
@iamprayush I see you posted screenshots of 3 passes of the extensions suite, but since the extensions utilities can be used by other e2e tests, we need to rerun all the tests, both on TravisCI and on CircleCI, not just the extensions suite. This policy is described in more detail on the [e2e wiki page](https://github.com/oppia/oppia/wiki/End-to-End-Tests#temporary-flakiness-mitigation-measures)

U8NWXD commented:
@iamprayush alternatively, you can post screenshots of 5 consecutive passes of *every* affected test suite. This might be a lot though since any suite that creates interactions is potentially affected

iamprayush committed:
Minor fix

seanlip commented on core/domain/exp_jobs_one_off.py:
```diff
@@ -293,32 +293,8 @@ def map(item):
-                    yield ('SUCCESS_WITH_OUTPUT', output_values.encode('utf-8'))
-                else:
-                    yield ('SUCCESS', None)
+                yield ('SUCCESS', None)
```
Typically we use ('SUCCESS', 1). (more semantic, for counting up the length). Please fix in a follow-up PR.

iamprayush replied:
Will do! Added it to this [issue](https://github.com/oppia/oppia/issues/10229).

iamprayush commented:
@U8NWXD I've added 3 successful runs of all tests on TravisCI and CircleCI, PTAL, thanks!

U8NWXD reviewed (approved):
Thanks for the screenshots @iamprayush! This looks great
